src/common/OpQueue.h

```diff
@@ -0,0 +1,65 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2004-2006 Sage Weil <sage@newdream.net>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#ifndef OP_QUEUE_H
+#define OP_QUEUE_H
+
+#include "common/Formatter.h"
+#include "include/msgr.h"
+
+#include <list>
+#include <functional>
+
+/**
+ * Abstract class for all Op Queues
+ *
+ * In order to provide optimized code, be sure to declare all
+ * virutal functions as final in the derived class.
+ */
+
+template <typename T, typename K>
+class OpQueue {
+
+  public:
+    // How many Ops are in the queue
+    virtual unsigned length() const = 0;
+    // Ops will be removed and placed in *removed if f is true
+    virtual void remove_by_filter(
+	std::function<bool (T)> f, std::list<T> *removed = 0) = 0;
+    // Ops of this priority should be deleted immediately
+    virtual void remove_by_class(K k, std::list<T> *out = 0) = 0;
+    // Enqueue the op for processing. If front is true then the
+    // op should be put in the front of the queue, otherwise it
+    // should be queued in the back. The class allows for multiple
+    // queues to be specfied with different properties. Currently
+    // there is a strict queue which should dequeue before normal
+    // ops for things like peering and maintenance.
+    virtual void enqueue(
+	K cl,
+	T item,
+	unsigned priority,
+	unsigned cost = 0,
+	bool front = CEPH_OP_QUEUE_BACK,
```

Comment:
I don't see any advantage to replacing enqueue[_strict][_front] with run time arguments.  strict and front are actually required by *every* implementation of this interface because OSD semantics depend on it.